kentikapi/errors/errors.go

```diff
@@ -0,0 +1,231 @@
+package kentik_errors
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type Code int
+
+const (
+	AuthError          Code = 0
+	InvalidRequest     Code = 1
+	NotFound           Code = 2
+	RateLimitExhausted Code = 3
+	Temporary          Code = 4
+	Timeout            Code = 5
+)
+
+type KentikError struct {
+	Codes []Code
+	Msg   string
+}
+
+func (ke KentikError) Error() string {
+	return fmt.Sprintf("category: %v, message: %s", ke.Codes, ke.Msg)
+}
+
+func (ke KentikError) GetKentikError() *KentikError { return &ke }
+
+func GetCodes(err error) ([]Code, bool) {
+	var tErr interface {
+		GetKentikError() *KentikError
+	}
+	if ok := errors.As(err, &tErr); ok {
+		ktError := tErr.GetKentikError()
+		return ktError.Codes, true
+	}
+	return []Code{}, false
+}
+
+func KentikErrorUpdateMsg(msg string, err error) error {
+	var tErr interface {
+		GetKentikError() *KentikError
+	}
+	if ok := errors.As(err, &tErr); ok {
+		ktError := tErr.GetKentikError()
+		ktError.Msg = msg
+		return ktError
+	}
+	return err
+}
+
+func KentikErrorFromHTTP(response *http.Response, err error) error {
+	if err == nil {
+		return nil
+	}
+	ke := KentikError{Msg: err.Error()}
+	if response != nil {
+		switch response.StatusCode {
+		case http.StatusUnauthorized, http.StatusForbidden:
+			ke.Codes = []Code{AuthError}
+		case http.StatusNotFound:
+			ke.Codes = []Code{NotFound}
+		case http.StatusRequestTimeout:
+			ke.Codes = []Code{Timeout, Temporary}
+		case http.StatusTooManyRequests:
+			ke.Codes = []Code{RateLimitExhausted, Temporary}
+		case http.StatusServiceUnavailable, http.StatusGatewayTimeout:
+			ke.Codes = []Code{Temporary}
+		default:
+			ke.Codes = []Code{InvalidRequest}
+		}
+	} else {
+		ke.Codes = []Code{Timeout}
+	}
+	return ke
+}
+
+func KentikErrorFromGRPC(err error) error {
+	if err == nil {
+		return nil
+	}
+	ke := KentikError{Msg: err.Error()}
+	if e, ok := status.FromError(err); ok {
+		switch e.Code() {
+		case codes.DeadlineExceeded:
+			ke.Codes = []Code{Timeout, Temporary}
+		case codes.NotFound:
+			ke.Codes = []Code{NotFound}
+		case codes.PermissionDenied, codes.Unauthenticated:
+			ke.Codes = []Code{AuthError}
+		case codes.ResourceExhausted:
+			ke.Codes = []Code{RateLimitExhausted, Temporary}
+		case codes.Unavailable:
+			ke.Codes = []Code{Temporary}
```

Comment:
```suggestion
			ke.Codes = []Code{Unavailable}
```
I know this is temporary error, but I'd rather keep root cause in error code, so that this info is not lost

Reply:
Done